New feature flag to keep the frame pointer for LLVM-generated functions.

mcourteaux opened the pull request:
I implemented this as a target feature flag. This way we can still disable this for actual performance optimized releases of software.

abadams commented:
Compiler flags in the cmakefile should be irrelevant here, if we do things right. There are other ways to build libHalide, so these should be controlled in code, not in the build. These flags should just set the initial state of the function attributes in the runtime functions, but they should get clobbered by CodeGen_LLVM.cpp line 629. 

These functions are currently marked nounwind because it's impossible to throw exceptions safely through Halide-generated code (we'd leak all over the place), and according to the llvm docs, nounwind is for functions you can't throw exceptions through. There's another attribute "uwtable" which indicates "the ABI being targeted requires that an unwind table entry be produced for this function even if we can show that no exceptions passes by it." We could try adding that, with fn.addFnAttr("uwtable") in CodeGen_Internal.cpp in set_function_attributes_from_halide_target_options.

slomp commented:
Yeah, and the unwind tables are a hack to do callstack unwinding in the case exception handling `.eh_frame` sections are available, which we can't rely on anyway.

@shoaibkamil I tried adding `-fno-omit-frame-pointer` to the runtime flags, but still no stack frames past the first runtime symbol that shows up. 
Not sure what else we may need to change to ensure we emit frame pointers for things generated by `Halide::compile_standalone_runtime()`.

mcourteaux commented:
> I just gave the fn.addFnAttr(\"frame-pointer\", \"all\") a try on Windows, and it did not give me what I was expecting.

We could add those flags for the runtime functions too. I'll try that tomorrow and push it. You can try again. I'll ping you when I did this.

abadams commented:
Working with Marcos, the thing that worked on windows was:

```
    fn.addFnAttr(llvm::Attribute::getWithUWTableKind(module.getContext(), llvm::UWTableKind::Default));
```

Turns out windows x64 doesn't really do frame pointers at all.

slomp commented:
> I'm sort of thinking that the Debug feature flag should enable the required flags to make it debuggable.

The thing is: callstacks are also invaluable for profiling, and profiling Debug modules is not ideal.

> What do we do with this flag?

The safe thing to do for now is for unwinding to be "**opt-in**".

We would need to do some research to see if the impact of adding `.eh_frame` sections for Halide code and runtime modules is either onerous or negligible. The same goes for frame-pointer (since this works on Linux, and probably on Mac too).

abadams commented:
My inclination would be to have a single flag that turns on frame pointers and unwind tables, and call it something to do with backtraces or call stacks. EnableBacktraces? PreserveCallStack?

mcourteaux commented:
Update: changed the flag name to `EnableBacktraces` and added the UWTable fn attribute, and additionally also enable this when Debug feature flag is set.

alexreinking commented:
That `correctness_register_shuffle` failure seems specific to LLVM 21, which we weren't testing until this weekend.

mcourteaux commented:
@slomp is this working for you, on Windows?

slomp commented:
Where do we all stand with this one?
I'm inclined in merging, we just need to all agree on the name/label of the feature flag.
(I'm OK with the current `"enable_backtraces"`.)

mcourteaux commented:
I'm also in favor of merging as is. Waiting for approval.

mcourteaux commented:
@abadams @alexreinking @slomp Approval to land? I'm juggling PRs.

abadams commented on src/runtime/HalideRuntime.h:
```diff
@@ -1355,10 +1355,11 @@ extern int halide_error_vscale_invalid(void *user_context, const char *func_name
-    halide_target_feature_no_bounds_query,  ///< Disable the bounds querying functionality.
+    halide_target_feature_jit = 0,            ///< Generate code that will run immediately inside the calling process.
+    halide_target_feature_debug,              ///< Turn on debug info and output for runtime code.
+    halide_target_feature_enable_backtraces,  ///< Keep the frame pointer in tact in functions produced by LLVM.
```
There should be something about unwind tables. Maybe "Preserve frame pointers and include unwind tables to support accurate backtraces for debugging and profiling."

abadams reviewed (approved):
looks good to me with a minor comment about the description of the feature flag

alexreinking commented:
@mcourteaux - go ahead and make the documentation tweak @abadams suggested and then just merge.

alexreinking commented on src/Target.h:
```diff
@@ -84,6 +84,7 @@ struct Target {
     enum Feature {
         JIT = halide_target_feature_jit,
         Debug = halide_target_feature_debug,
+        EnableBacktraces = halide_target_feature_enable_backtraces,
```
Ahhh @mcourteaux - I'm so sorry to block you so late, but can you please add this to PyEnums.cpp, too, and leave a comment in both source locations to cross-check them?

alexreinking replied:
I'll change to an approve when you ACK this

alexreinking reviewed (changes requested):
Needs a one-line change for Python 😭
